### DISCUSSION TITLE
What happens when multiple packageRules match? How the individual config elements from these matches are merged?

### DISCUSSION THREAD

**Discussion by phaninra** [renovatebot/renovate#18437] [Request Help] [answered]
I read from the other posts that the last matching rule wins. However there's more to it. Here's an example

```
module.exports = {

    enabledManagers: ["gradle-wrapper"],

    addLabels: ["label-1"],

    packageRules: [

        {
            matchManagers: ["gradle-wrapper"],
            addLabels: ["packageRule-1", "updateType: {{updateType}}"],
        },

        {
            addLabels: ["packageRule-2"],
        },

        {
            matchManagers: ["gradle-wrapper"],
            allowedVersions: ["<=7.5.1"],
            addLabels: ["packageRule-3"],
        },

        {
            matchManagers: ["gradle-wrapper"],
            allowedVersions: ["<=7.4.2"],
            addLabels: ["packageRule-4"],
        },
    ],
}
```

Outcome Pull Request
- Proposes the update to "7.4.2" which makes sense since that is the last rule that matches
- Labels include "label-1", "packageRule-1", "packageRule-3", "packageRule-4", "updateType: minor"

I'm trying to understand 
- Why the labels include "packageRule-3"
- Why the labels include "packageRule-1"
- Why the labels don't include "packageRule-2"
- How is the "{{updateType}}" in the label evaluated when it is in the context of rule-1 and the actual update happens in the context of final rule?

So how are multiple matching rules dealt with? What happens to other config elements besides addLabels?

**Comment by viceice** [comment, marked as answer]
all matching rules are applied and addLabels is mergable, so all matching labels are applied.

checkout mergable flag for options in docs

**Reply by phaninra** [reply]
Thank you, that makes sense.  Once all matching rules are discovered for a package,  all config elements are merged into a single rule. Is that right?

**Reply by viceice** [reply]
they are applied in order to a resulting config. it depends on the mergable config flag if options are merged (mostly array options only) or if the next rule override the previous.
